### PR TITLE
qutebrowser: add onChange ipc reloading

### PR DESCRIPTION
**Pull request by Misterio77**
### Description

Add a `onChange` entry for qutebrowser, allowing it to react to configuration changes.

Qutebrowser can reload its configuration with the `:config-source` command, which can be run through its IPC socket too.

### Checklist

- [x] Change is backwards compatible.

- [x] Code formatted with `./format`.

- [x] Code tested through `nix-shell --pure tests -A run.all`.

- [ ] Test cases updated/added. See [example](https://github.com/nix-community/home-manager/commit/f3fbb50b68df20da47f9b0def5607857fcc0d021#diff-b61a6d542f9036550ba9c401c80f00ef).

- [x] Commit messages are formatted like

    ```
    {component}: {description}

    {long description}
    ```

    See [CONTRIBUTING](https://github.com/nix-community/home-manager/blob/master/doc/contributing.adoc#sec-commit-style) for more information and [recent commit messages](https://github.com/nix-community/home-manager/commits/master) for examples.

Here's how it looks:

https://user-images.githubusercontent.com/5727578/125020477-49cad580-e04f-11eb-8d61-0ea1a1cb333e.mp4
    
    
I was using something similar to [this script](https://github.com/qutebrowser/qutebrowser/blob/master/scripts/open_url_in_instance.sh) to update my qutebrowser instance when changing themes. I noticed home-manager [automatically reloads sway](https://github.com/nix-community/home-manager/blob/master/modules/services/window-managers/i3-sway/sway.nix#L435) when needed, so i tried to get it working for qutebrowser too.

A few points:
- I'm not sure if tests can even be written for something like this, but i'd be glad to if someone pointed me in the right direction.
- I'm also not sure whether MacOS defines these variables, so i'm not sure if it runs on anything other than Linux, so i added it specifically to the entry inside `hostPlatform.isLinux`. If anyone reading has a Mac, please feel free to paste that into the `hostPlatform.isDarwin` section and try it out.
- Is `${pkgs.coreutils}` needed for stuff such as `echo`? I thought it was, so used it that way. Open to suggestions if that's not the usual way to do it.

I have something of a itch to dynamically theme the entire system through home-manager, so i'll probably open PRs for other programs as i figure them out.

### CODE REVIEW

File: modules/programs/qutebrowser.nix [hunk 1]
```diff
@@ -306,7 +306,17 @@ in {
       };
 
     xdg.configFile."qutebrowser/config.py" =
-      mkIf pkgs.stdenv.hostPlatform.isLinux { text = qutebrowserConfig; };
+      mkIf pkgs.stdenv.hostPlatform.isLinux {
+        text = qutebrowserConfig;
+        onChange = ''
+          hash="$(${pkgs.coreutils}/bin/echo -n $USER | ${pkgs.coreutils}/bin/md5sum | ${pkgs.coreutils}/bin/cut -d' ' -f1)"
```
Comment (berbiche) [inline modules/programs/qutebrowser.nix:312]:
```suggestion
          hash="$(echo -n \"$USER\" | ${pkgs.coreutils}/bin/md5sum | ${pkgs.coreutils}/bin/cut -d' ' -f1)"
```

The default shell used in scripts is defined by `pkgs.stdenv.shell` (bash globally).
AFAIK changing the default shell will incur a global rebuild.
This doesn't mean we can use and rely on bashism.

On a second note, popular shells like zsh, ksh, bash and fish all implement echo as a builtin command.

File: modules/programs/qutebrowser.nix [hunk 2]
```diff
@@ -306,7 +306,17 @@ in {
...
+          socket="''${XDG_RUNTIME_DIR:-/run/user/$UID}/qutebrowser/ipc-$hash"
+          if [ -S $socket ]; then
+            ${pkgs.coreutils}/bin/echo "Reloading qutebrowser"
```
Comment (berbiche) [inline modules/programs/qutebrowser.nix:315]:
```suggestion
            echo "Reloading qutebrowser"
```

File: modules/programs/qutebrowser.nix [hunk 3]
```diff
@@ -306,7 +306,17 @@ in {
...
+            $DRY_RUN_CMD ${pkgs.coreutils}/bin/printf '{"args": [":config-source"], "target_arg": null, "protocol_version": 1}\n' | ${pkgs.socat}/bin/socat -lf /dev/null - UNIX-CONNECT:"''${socket}"
```
Comment (berbiche) [inline modules/programs/qutebrowser.nix:316]:
```suggestion
            $DRY_RUN_CMD echo '{"args": [":config-source"], "target_arg": null, "protocol_version": 1}' | ${pkgs.socat}/bin/socat -lf /dev/null - UNIX-CONNECT:"''${socket}"
```

`printf` is also a shell builtin in many shells.

You can optionally replace the json with `escapeShellArg (builtins.toJSON { args = [ ":config-source" ]; target_arg = null; protocol_version = 1 })`.
Just thought you'd  be interested in this neat trick!

**Review by berbiche — CHANGES_REQUESTED** [review]
Thanks!

**Comment by Misterio77** [comment]
@berbiche, thanks a lot for the review and suggestions.

It seems `$USER` has to be unquoted, or else i get the wrong md5 hash. Just fixed this in the last commit

I got `toJSON` working, but it seems the JSON keys and strings don't get quoted, unless i escape a double quote inside the already existing quote (quite ugly), and qutebrowser won't accept the command if they're not.

I wonder if there's any better way to do it?
```bash
command=${builtins.toJSON { "\"args\"" = [ "\":config-source\"" ]; "\"target_arg\"" = null; "\"protocol_version\"" = 1; }}
$DRY_RUN_CMD echo $command | ${pkgs.socat}/bin/socat -lf /dev/null - UNIX-CONNECT:$socket
```

**Comment by Misterio77** [comment]
Oooh that's what the `escapeShellArg` is for, got it working now.

File: modules/programs/qutebrowser.nix [hunk 4]
```diff
@@ -306,7 +306,18 @@ in {
       };
 
     xdg.configFile."qutebrowser/config.py" =
-      mkIf pkgs.stdenv.hostPlatform.isLinux { text = qutebrowserConfig; };
+      mkIf pkgs.stdenv.hostPlatform.isLinux {
+        text = qutebrowserConfig;
+        onChange = ''
+          hash="$(echo -n $USER | ${pkgs.coreutils}/bin/md5sum | ${pkgs.coreutils}/bin/cut -d' ' -f1)"
```
Comment (rycee) [inline modules/programs/qutebrowser.nix:312]: The coreutils package is in path:

```suggestion
          hash="$(echo -n $USER | md5sum | cut -d' ' -f1)"
```

File: modules/programs/qutebrowser.nix [hunk 5]
```diff
@@ -306,7 +306,18 @@ in {
...
+          socket="''${XDG_RUNTIME_DIR:-/run/user/$UID}/qutebrowser/ipc-$hash"
+          if [ -S $socket ]; then
+            echo "Reloading qutebrowser"
+            command=${escapeShellArg (builtins.toJSON { args = [ ":config-source" ]; target_arg = null; protocol_version = 1; })}
+            $DRY_RUN_CMD echo $command | ${pkgs.socat}/bin/socat -lf /dev/null - UNIX-CONNECT:$socket
+          fi
+        '';
```
Comment (rycee) [inline modules/programs/qutebrowser.nix:319]: When introducing variables in an activation block then it is best to make sure they are removed after use:
```suggestion
          unset hash socket command
        '';
```

**Comment by rycee** [comment]
Looks good I think, just some minor comments. Btw, you'll need to run the format script in the project root.

**Comment by Misterio77** [comment]
Thanks for the help, @rycee!

Should i squash my commits?

**Comment by rycee** [comment]
@Misterio77 No need, I've squashed and merged now :-)

**Pull request merged**